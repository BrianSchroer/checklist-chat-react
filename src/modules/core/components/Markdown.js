import React from 'react';
import PropTypes from 'prop-types';
import ReactMarkdown from 'react-markdown';
import breaks from 'remark-breaks';

const Markdown = ({ source, className }) => {
  return (
    <ReactMarkdown
      plugins={[breaks]}
      className={className}
      source={source}
      disallowedTypes={['heading', 'thematicBreak']}
      escapeHtml
      unwrapDisallowed={false}
    />
  );
};

Markdown.propTypes = {
  source: PropTypes.string.isRequired,
  className: PropTypes.string
};

export default Markdown;
