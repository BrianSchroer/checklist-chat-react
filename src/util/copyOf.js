export default original => Object.assign({}, original);
