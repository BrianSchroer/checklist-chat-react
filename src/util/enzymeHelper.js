const enzymeHelper = class {

    static nestedFind(wrapper, selector) {
        const selectors = selector.split('>').map(node => node.trim());
        let parent = wrapper;
        let found = [];

        for (let i = 0; i < selectors.length; i++) {
            found = parent.find(selectors[i]);
            if (!found.length) {
                break;
            }
        }

        return found;
    }

    static findSingle(wrapper, selector) {
        const findArray = enzymeHelper.nestedFind(wrapper, selector);

        if (findArray.length != 1) {
            throw(`expected 1 find for selector "${selector}", but found ${findArray.length}.`);
        }

        return findArray.first();
    }
};

export default enzymeHelper;
