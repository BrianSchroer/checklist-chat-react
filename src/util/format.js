import moment from 'moment';

export default class {
    static time(input = '') {

        const isoDateFormatRegex = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}([+-]\d{2}:\d{2})?/g;
        let formatted = input;

        if (input && input.match(isoDateFormatRegex)) {
            formatted = this.dateTimeString(input, 'h:mm a');
        }

        return formatted;
    }

    static dateTimeString(input, formatString) {
        let formatted = input;

        if (input) {
            const testMoment = moment(input);

            if (testMoment.isValid()) {
                const testFormat = testMoment.format(formatString);
                if (testFormat != 'Invalid date') {
                    formatted = testFormat;
                }
            }
        }

        return formatted;
    }
}
