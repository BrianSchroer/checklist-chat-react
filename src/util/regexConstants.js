export const CSS_PATH = /(\.css)$/;
export const EOT_PATH = /\.eot(\?v=\d+\.\d+\.\d+)?$/;
export const JS_PATH = /\.js$/;
export const SVG_PATH = /\.svg(\?v=\d+\.\d+\.\d+)?$/;
export const TTF_PATH = /\.ttf(\?v=\d+\.\d+\.\d+)?$/;
export const WOFF_OR_WOFF2_PATH = /\.(woff|woff2)(\?v=\d+\.\d+\.\d+)?$/;
