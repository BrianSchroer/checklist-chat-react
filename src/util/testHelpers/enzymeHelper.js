import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

export default class EnzymeHelper {
  /**
   * Create new EnzymeHelper instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propsAdjustments) =>
      Object.assign({}, props, propsAdjustments);
    this.wrapper = null;
  }

  /**
   * Fluent syntax helper to specify a callback function to be called by the test() function
   * to assign propsAdjustments to the props passed that were passed via the constructor.
   * (Defaults to assigning top-level props only.)
   */
  withPropsAdjuster = propsAdjuster => {
    this.adjustProps = propsAdjuster;
    return this;
  };

  /**
   * Creates a new props object by calling the propsAdjuster function to apply
   * propsAdjustments (if any) to the props passed to the constructor, then
   * returns a "shallowWrapper" of the component with the specified props.
   */
  shallow = propsAdjustments => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propsAdjustments);

    return (this.wrapper = shallow(<Component {...props} />));
  };

  /**
   * This method wraps the enzyme find method with the ability to perform
   * "nested" finds (e.g. "div.main > div.sub > p.instructions") for the shallow "wrapper"
   * created/returned by a previous call to the "shallow" function.
   *
   * @static
   * @param {any} selector for the search
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   * @returns array of found results
   */
  find = (selector, higherOrderComponent) => {
    return this.findIn(this.wrapper, selector, higherOrderComponent);
  };

  /**
   * This method wraps the enzyme find method with the ability to perform
   * "nested" finds (e.g. "div.main > div.sub > p.instructions").
   *
   * @param {any} wrapper to be searched
   * @param {any} selector for the search
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   * @returns array of found results
   */
  findIn = (wrapper, selector, higherOrderComponent) => {
    const selectors = selector.split('>').map(node => node.trim());
    let parent = wrapper;
    let found = [];

    for (let i = 0; i < selectors.length; i++) {
      found = parent.find(selectors[i]);

      if (found.length == 0) {
        if (higherOrderComponent) {
          found = parent.find(`${higherOrderComponent}(${selectors[i]})`);
        }

        if (found.length == 0) {
          break;
        }
      }
    }

    return found;
  };

  /**
   * Call "find" function and assert the expected match count.
   *
   * @param {number} expectedCount
   * @param {any} selector for the search (can be "nested"; e.g. "div > span")
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   * @returns array of found results
   */
  assertFindCount = (expectedCount, selector, higherOrderComponent) => {
    return this.assertFindCountIn(
      this.wrapper,
      expectedCount,
      selector,
      higherOrderComponent
    );
  };

  /**
   * Call "findIn" function and assert the expected match count.
   *
   * @param {any} wrapper to be searched
   * @param {number} expectedCount
   * @param {any} selector for the search (can be "nested"; e.g. "div > span")
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   * @returns array of found results
   */
  assertFindCountIn = (
    wrapper,
    expectedCount,
    selector,
    higherOrderComponent
  ) => {
    const found = this.findIn(wrapper, selector, higherOrderComponent);

    if (found.length != expectedCount) {
      const errorMessage = [
        `Expected ${expectedCount} element(s) for selector "${selector}"` +
          `, but found ${found.length}.`,
        '',
        this.getDebugHtml(wrapper),
        ''
      ].join('\n');

      throw new Error(errorMessage);
    }

    return found;
  };

  /**
   * Call "find" method and assert that no matches are found.
   *
   * @param {any} selector for the search (can be "nested"; e.g. "div > span")
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   */
  assertNoMatch = (selector, higherOrderComponent) => {
    this.assertNoMatchIn(this.wrapper, selector, higherOrderComponent);
  };

  /**
   * Call "findIn" method and assert that no matches are found.
   *
   * @param {any} wrapper to be searched
   * @param {any} selector for the search (can be "nested"; e.g. "div > span")
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   */
  assertNoMatchIn = (wrapper, selector, higherOrderComponent) => {
    this.assertFindCountIn(wrapper, 0, selector, higherOrderComponent);
  };

  /**
   * Call "find" function and assert that one and only one match is found.
   *
   * @param {any} selector for the search (can be "nested"; e.g. "div > span")
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   * @returns the found node
   */
  findSingle = (selector, higherOrderComponent) => {
    return this.findSingleIn(this.wrapper, selector, higherOrderComponent);
  };

  /**
   * Call "findIn" function and assert that one and only one match is found.
   *
   * @param {any} wrapper to be searched
   * @param {any} selector for the search (can be "nested"; e.g. "div > span")
   * @param {string} higherOrderComponent = optional higher-order component (e.g. "Connect")
   *      that selector may be wrapped with.
   * @returns the found node
   */
  findSingleIn = (wrapper, selector, higherOrderComponent) => {
    const found = this.assertFindCountIn(
      wrapper,
      1,
      selector,
      higherOrderComponent
    );

    return found.first();
  };

  /**
   * Returns an HTML-like string of the shallow wrapper for debugging purposes. Useful to
   * log to the console when tests are not passing when you expect them to.
   *
   * @param {any} wrapper (if not passed, uses the wrapper retrieved/returned by preceding "shallow" function call)
   * @returns the resulting string
   */
  getDebugHtml = wrapper => (wrapper || this.wrapper).debug();

  /**
   * Logs an HTML-like string of the wrapper for debugging purposes. Useful to
   * log to the console when tests are not passing when you expect them to.
   *
   * @param {any} wrapper (if not passed, uses the wrapper retrieved/returned by preceding "shallow" function call)
   * @returns the resulting string
   */
  logDebugHtml = wrapper =>
    console.log(this.getDebugHtml(wrapper || this.wrapper)); // eslint-disable-line no-console
}
