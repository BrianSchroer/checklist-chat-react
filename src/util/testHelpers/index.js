import {shallow, mount} from 'enzyme';
import enzymeHelper from './enzymeHelper';

export {
    shallow as shallow,
    mount as mount,
    enzymeHelper as enzymeHelper
};
