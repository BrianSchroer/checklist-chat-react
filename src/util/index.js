import format from './format';
import uiHelper from './uiHelper';
import validationHelper from './validationHelper';

export {
    format as format,
    uiHelper as uiHelper,
    validationHelper as validationHelper
};
