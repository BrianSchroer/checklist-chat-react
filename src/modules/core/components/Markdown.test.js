import React from 'react';
import { snapshotHelper } from '../../../util/testHelpers';
import Markdown from './Markdown';

function assertSnapshotMatch(...sourceStrings) {
  assertSnapshotMatchWithClassName(sourceStrings.join(''));
}

function assertSnapshotMatchWithClassName(source, className) {
  snapshotHelper.assertMatch(
    <div>
      <h4>source:</h4>
      <pre>{source}</pre>
      <h4>className:</h4>
      <div>{className || '(none)'}</div>
      <h4>output:</h4>
      <Markdown source={source} className={className} />
    </div>
  );
}

describe('Markdown', () => {
  it('should render source without markdown', () =>
    assertSnapshotMatch(
      'Source without any markdown formatting renders as a "p" within a "div".'
    ));

  it('should not allow headings', () =>
    assertSnapshotMatch('# H1', '\n## H2', '\n### H3'));

  it('should render source with className', () =>
    assertSnapshotMatchWithClassName(
      'This should be italic.',
      'chat-action-message-text'
    ));

  it('should render emphasis (italic)', () =>
    assertSnapshotMatch(
      'Text within *asterisks* or _underscores_ is rendered within "em" tags.'
    ));

  it('should render strong (bold)', () =>
    assertSnapshotMatch(
      'Text within **double asterisks** or **double underscores**',
      ' is rendered within "strong" tags.'
    ));

  it('should render links', () =>
    assertSnapshotMatch(
      'Here are links to the "react-markdown" github repo:',
      '\n[https://github.com/rexxars/react-markdown](https://github.com/rexxars/react-markdown),',
      '\nand to  a [markdown reference page](http://commonmark.org/help/)'
    ));

  it('should render images', () =>
    assertSnapshotMatch(
      "Here's an image: ",
      '![Image](https://media.giphy.com/media/OxgjxSpRwMXoA/giphy.gif)'
    ));

  it('should render line feeds', () =>
    assertSnapshotMatch('Line1 \nLine2 \nLine3 \n(\\n is converted to <br>)'));

  it('should render unordered lists', () =>
    assertSnapshotMatch(
      'can be specified with asterisks:',
      '\n* item 1',
      '\n* item 2',
      '\n\n...or with dashes:',
      '\n- item 1',
      '\n- item 2'
    ));

  it('should render ordered lists', () =>
    assertSnapshotMatch(
      'can be specified like this:',
      '\n1. item 1',
      '\n2. item 2',
      '\n\n...or like this:',
      '\n1) item 1',
      '\n2) item 2'
    ));

  it('should not render horizontal rules', () =>
    assertSnapshotMatch(
      'Horizontal rule cannot be specified with three underscores:',
      '\n___',
      '\nor with three asterisks',
      '\n***'
    ));

  it('should render inline code', () =>
    assertSnapshotMatch(
      'An arrow function looks like this: `number => number + 1`.'
    ));

  it('should render code blocks', () =>
    assertSnapshotMatch(
      "Here's some assertSnapshotMatch JavaScript:",
      '\n```',
      '\nfunction addOne(number) {',
      '\n    return number + 1;',
      '\n}',
      '\n```',
      '\nThat was a code block.'
    ));

  it('should render block quotes', () =>
    assertSnapshotMatch(
      "Hofstadter's law states:",
      "\n> It always takes longer than you expect, even when you take into account Hofstadter's Law",
      "\n\nThat was Hofstadter's law."
    ));

  it('should escape inline HTML', () =>
    assertSnapshotMatch(
      'Inline HTML gets "escaped", like this: <span>inline HTML</span>'
    ));
});
