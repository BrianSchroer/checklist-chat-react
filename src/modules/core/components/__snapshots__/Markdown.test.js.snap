// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown should escape inline HTML 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Inline HTML gets "escaped", like this: &lt;span&gt;inline HTML&lt;/span&gt;
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Inline HTML gets "escaped", like this: 
      <span>
        &lt;span&gt;
      </span>
      inline HTML
      <span>
        &lt;/span&gt;
      </span>
    </p>
  </div>
</div>
`;

exports[`Markdown should not allow headings 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    # H1
## H2
### H3
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div />
</div>
`;

exports[`Markdown should not render horizontal rules 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Horizontal rule cannot be specified with three underscores:
___
or with three asterisks
***
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Horizontal rule cannot be specified with three underscores:
    </p>
    <p>
      or with three asterisks
    </p>
  </div>
</div>
`;

exports[`Markdown should render block quotes 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Hofstadter's law states:
&gt; It always takes longer than you expect, even when you take into account Hofstadter's Law

That was Hofstadter's law.
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Hofstadter's law states:
    </p>
    <blockquote>
      <p>
        It always takes longer than you expect, even when you take into account Hofstadter's Law
      </p>
    </blockquote>
    <p>
      That was Hofstadter's law.
    </p>
  </div>
</div>
`;

exports[`Markdown should render code blocks 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Here's some assertSnapshotMatch JavaScript:
\`\`\`
function addOne(number) {
    return number + 1;
}
\`\`\`
That was a code block.
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Here's some assertSnapshotMatch JavaScript:
    </p>
    <pre>
      <code>
        function addOne(number) {
    return number + 1;
}
      </code>
    </pre>
    <p>
      That was a code block.
    </p>
  </div>
</div>
`;

exports[`Markdown should render emphasis (italic) 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Text within *asterisks* or _underscores_ is rendered within "em" tags.
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Text within 
      <em>
        asterisks
      </em>
       or 
      <em>
        underscores
      </em>
       is rendered within "em" tags.
    </p>
  </div>
</div>
`;

exports[`Markdown should render images 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Here's an image: ![Image](https://media.giphy.com/media/OxgjxSpRwMXoA/giphy.gif)
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Here's an image: 
      <img
        alt="Image"
        src="https://media.giphy.com/media/OxgjxSpRwMXoA/giphy.gif"
      />
    </p>
  </div>
</div>
`;

exports[`Markdown should render inline code 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    An arrow function looks like this: \`number =&gt; number + 1\`.
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      An arrow function looks like this: 
      <code>
        number =&gt; number + 1
      </code>
      .
    </p>
  </div>
</div>
`;

exports[`Markdown should render line feeds 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Line1 
Line2 
Line3 
(\\n is converted to &lt;br&gt;)
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Line1
      <br />
      Line2
      <br />
      Line3
      <br />
      (\\n is converted to 
      <span>
        &lt;br&gt;
      </span>
      )
    </p>
  </div>
</div>
`;

exports[`Markdown should render links 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Here are links to the "react-markdown" github repo:
[https://github.com/rexxars/react-markdown](https://github.com/rexxars/react-markdown),
and to  a [markdown reference page](http://commonmark.org/help/)
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Here are links to the "react-markdown" github repo:
      <br />
      <a
        href="https://github.com/rexxars/react-markdown"
      >
        https://github.com/rexxars/react-markdown
      </a>
      ,
      <br />
      and to  a 
      <a
        href="http://commonmark.org/help/"
      >
        markdown reference page
      </a>
    </p>
  </div>
</div>
`;

exports[`Markdown should render ordered lists 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    can be specified like this:
1. item 1
2. item 2

...or like this:
1) item 1
2) item 2
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      can be specified like this:
      <br />
      1. item 1
      <br />
      2. item 2
    </p>
    <p>
      ...or like this:
      <br />
      1) item 1
      <br />
      2) item 2
    </p>
  </div>
</div>
`;

exports[`Markdown should render source with className 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    This should be italic.
  </pre>
  <h4>
    className:
  </h4>
  <div>
    chat-action-message-text
  </div>
  <h4>
    output:
  </h4>
  <div
    className="chat-action-message-text"
  >
    <p>
      This should be italic.
    </p>
  </div>
</div>
`;

exports[`Markdown should render source without markdown 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Source without any markdown formatting renders as a "p" within a "div".
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Source without any markdown formatting renders as a "p" within a "div".
    </p>
  </div>
</div>
`;

exports[`Markdown should render strong (bold) 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    Text within **double asterisks** or **double underscores** is rendered within "strong" tags.
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      Text within 
      <strong>
        double asterisks
      </strong>
       or 
      <strong>
        double underscores
      </strong>
       is rendered within "strong" tags.
    </p>
  </div>
</div>
`;

exports[`Markdown should render unordered lists 1`] = `
<div>
  <h4>
    source:
  </h4>
  <pre>
    can be specified with asterisks:
* item 1
* item 2

...or with dashes:
- item 1
- item 2
  </pre>
  <h4>
    className:
  </h4>
  <div>
    (none)
  </div>
  <h4>
    output:
  </h4>
  <div>
    <p>
      can be specified with asterisks:
    </p>
    <ul>
      <li>
        item 1
      </li>
      <li>
        item 2
      </li>
    </ul>
    <p>
      ...or with dashes:
    </p>
    <ul>
      <li>
        item 1
      </li>
      <li>
        item 2
      </li>
    </ul>
  </div>
</div>
`;
